crates/circuit/src/interner.rs

```diff
@@ -181,26 +414,138 @@ where
+    /// most sense for them to be the same.
+    pub fn merge_map_using<S>(
+        &mut self,
+        other: &Interner<S>,
+        mut map_fn: impl FnMut(&S) -> Option<Cow<T>>,
+        target: &mut InternedMap<S, T>,
+    ) where
+        S: ?Sized + ToOwned + Hash + Eq,
+    {
+        target.map.clear();
+        target.map.reserve(other.0.len());
+        for key in other.0.iter() {
+            target
+                .map
+                .push(map_fn(key.borrow()).map(|cow| self.insert_cow(cow)));
+        }
+    }
+
+    /// Merge another interner into this one.
+    ///
+    /// The output mapping converts [Interned] indices from `other` to their new representations in
+    /// `self`.  Strictly, the interners can be for different types, but in practice it likely makes
+    /// most sense for them to be the same.
+    pub fn merge_map<S>(
+        &mut self,
+        other: &Interner<S>,
+        map_fn: impl FnMut(&S) -> Option<Cow<T>>,
+    ) -> InternedMap<S, T>
+    where
+        S: ?Sized + ToOwned + Hash + Eq,
+    {
+        let mut out = InternedMap::new();
+        self.merge_map_using(other, map_fn, &mut out);
+        out
+    }
+}
+
+impl<T> Interner<[T]>
+where
+    T: Hash + Eq + Clone,
+{
+    /// Merge another interner into this one, re-using the output storage for the key mapping.
+    ///
+    /// The mapping function is for scalar elements of the slice, as opposed to in [merge_map] where
+    /// it is for the entire key at once.  This function makes it easier to avoid allocations when
+    /// mapping slice-based conversions (though if `T` is not [Copy] and you're expecting there to
+    /// be a lot of true insertions during the merge, there is a potential clone inefficiency).
+    ///
+    /// If the `scalar_map_fn` returns `None` for any element of a slice, that entire slice is
+    /// filtered out from the merge.  The subsequent [InternedMap] will panic if the corresponding
+    /// [Interned] key is used as a lookup.
+    pub fn merge_map_slice_using<const N: usize>(
+        &mut self,
+        // Actually, `other` could be [Interner<[S]>], but then you'd need to specify `S` whenever
+        // you want to set `N`, which is just an API annoyance since we'll probably never need the
+        // two interners to be different types.
+        other: &Self,
+        mut scalar_map_fn: impl FnMut(&T) -> Option<T>,
+        target: &mut InternedMap<[T]>,
+    ) {
+        // Workspace for the mapping function. The aim here is that we're working on the stack, so
+        // the mapping doesn't need to make heap allocations.  We could either guess (which the
+        // higher-level `merge_slice_map` does), or force the caller to tell us how much stack space
+        // to allocate.  This method is lower leveel, so in this case we ask them to tell us; if
+        // they're optimising to the point of re-using the return allocations, they probably have a
```

Comment:
Hate to be doing this but...
```suggestion
        // they're optimizing to the point of re-using the return allocations, they probably have a
```
🇺🇸